src/jemalloc.c

```diff
@@ -4132,9 +4115,63 @@ batch_alloc(void **ptrs, size_t num, size_t size, int flags) {
 			batch_alloc_prof_sample_assert(tsd, batch, usize);
 		}
 
-		size_t n = arena_fill_small_fresh(tsd_tsdn(tsd), arena,
-		    ind, ptrs + filled, batch, zero);
-		filled += n;
+		size_t progress = 0;
+
+		if (likely(ind < SC_NBINS) && batch >= nregs) {
+			if (arena == NULL) {
+				unsigned arena_ind = mallocx_arena_get(flags);
+				if (arena_get_from_ind(tsd, arena_ind,
+				    &arena)) {
+					goto label_done;
+				}
+				if (arena == NULL) {
+					arena = arena_choose(tsd, NULL);
+				}
+				if (unlikely(arena == NULL)) {
+					goto label_done;
+				}
+			}
+			size_t arena_batch = batch - batch % nregs;
+			size_t n = arena_fill_small_fresh(tsd_tsdn(tsd), arena,
+			    ind, ptrs + filled, arena_batch, zero);
+			progress += n;
+			filled += n;
+		}
+
+		if (likely(ind < nhbins) && progress < batch) {
+			if (bin == NULL) {
+				unsigned tcache_ind = mallocx_tcache_get(flags);
+				tcache_t *tcache = tcache_get_from_ind(tsd,
+				    tcache_ind, /* slow */ true,
+				    /* is_alloc */ true);
+				if (tcache != NULL) {
+					bin = &tcache->bins[ind];
+				}
+			}
+			if (bin != NULL) {
+				size_t bin_batch = batch - progress;
+				size_t n = cache_bin_alloc_batch(bin, bin_batch,
+				    ptrs + filled);
+				if (config_stats) {
+					bin->tstats.nrequests += n;
+				}
+				if (zero) {
+					for (size_t i = 0; i < n; ++i) {
+						memset(ptrs[filled + i], 0,
+						    usize);
+					}
+				}
+				if (config_prof && opt_prof
+				    && unlikely(ind >= SC_NBINS)) {
+					for (size_t i = 0; i < n; ++i) {
+						prof_tctx_reset_sampled(tsd,
+						    ptrs[filled + i]);
+					}
+				}
+				progress += n;
+				filled += n;
+			}
+		}
 
```

Comment:
It seems like this will never allocate out of the arena bins -- is that intentional?

Reply:
No, it first allocates from the arena bins, but only in exact multiples of `nregs`. See the `if (likely(ind < SC_NBINS) && batch >= nregs)` branch. You can also see in the unit tests that I'm testing locality for the segment `[0, batch / nregs * nregs)`. I understand that the logic is a bit confusing here though; the `while` loop gets overly long in this revision - I'll try to do some cleanup.

Reply:
Sorry - forget about what I just said - I thought you meant fresh slabs...

Reply:
The allocation out of arena bins is done implicitly. When the cache bin is not enough, the `if (progress < batch || prof_sample_event)` branch at the end of the `while` loop serves to fill the cache bin, and then in the next round of the `while` loop, the batch gets allocations from the cache bin. There might be some unnecessary steps here, compared to directly filling in from arena bins whenever the cache bin is not enough, but I believe the overhead should not be a big deal: we at most try to fill `batch % nregs` allocations from the cache, and the cache has a size of twice of `nregs` (with the condition that it's between 20 and 200), so I'd imagine there's once or twice of filling in the worst case, and perhaps close to no filling at all in the average case. I didn't do very careful computation though, so may be wrong. I can change the logic if we figure that the overhead is not ideal.

Reply:
Ahhh I see, that's tricky. Could you add a comment explaining?